msal/tests/Test.MSAL.NET.Integration/UnifiedCacheTests.cs

```diff
@@ -0,0 +1,491 @@
+                };
+            }
+            msal::Microsoft.Identity.Client.TokenCacheExtensions.DeserializeUnifiedAndAdalCache(args.TokenCache, cacheData);
+        }
+
+        private static void MsalDoAfter(msal::Microsoft.Identity.Client.TokenCacheNotificationArgs args)
+        {
+            if (args.TokenCache.HasStateChanged)
+            {
+                msal::Microsoft.Identity.Core.Cache.CacheData cacheData =
+                    msal::Microsoft.Identity.Client.TokenCacheExtensions.SerializeUnifiedAndAdalCache(args.TokenCache);
+
+                lock (cacheStorageLock)
+                {
+                    AdalV3StateStorage = cacheData.AdalV3State;
+                    UnifiedStateStorage = cacheData.UnifiedState;
+                }
+            }
+        }
+
+        private void ClearCacheStorage()
+        {
+            AdalV3StateStorage = new byte[0];
+            UnifiedStateStorage = new byte[0];
+        }
+
+        private adal::Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext adalContext;
+        private adal::Microsoft.IdentityModel.Clients.ActiveDirectory.TokenCache adalCache;
+        private adal::Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationResult adalAuthResult;
+        private void InitAdal()
+        {
+            adalCache = new adal::Microsoft.IdentityModel.Clients.ActiveDirectory.TokenCache()
+            {
+                BeforeAccess = AdalDoBefore,
+                AfterAccess = AdalDoAfter
+            };
+            adalContext = new adal::Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext(Authority, adalCache);
+        }
+
+        private msal::Microsoft.Identity.Client.PublicClientApplication msalPublicClient;
+        private msal::Microsoft.Identity.Client.TokenCache msalCache;
+        private msal::Microsoft.Identity.Client.AuthenticationResult msalAuthResult;
+        private void InitMsal()
+        {
+            msalCache = new msal::Microsoft.Identity.Client.TokenCache();
+
+            msal::Microsoft.Identity.Client.TokenCacheExtensions.SetBeforeAccess(msalCache, MsalDoBefore);
+            msal::Microsoft.Identity.Client.TokenCacheExtensions.SetAfterAccess(msalCache, MsalDoAfter);
+
+            msalPublicClient = new msal::Microsoft.Identity.Client.PublicClientApplication(ClientId, Authority, msalCache);
+        }
+
+        private void ValidateMsalAuthResult()
+        {
+            Assert.IsNotNull(msalAuthResult);
+            Assert.IsNotNull(msalAuthResult.AccessToken);
+            Assert.IsNotNull(msalAuthResult.IdToken);
+            Assert.AreEqual(user.Upn, msalAuthResult.Account.Username);
+        }
+
+        private void ValidateAdalAuthResult()
+        {
+            Assert.IsNotNull(adalAuthResult);
+            Assert.IsNotNull(adalAuthResult.AccessToken);
+            Assert.IsNotNull(adalAuthResult.IdToken);
+            Assert.AreEqual(user.Upn, adalAuthResult.UserInfo.DisplayableId);
+        }
```

Comment:
should specific values be validated not just that something exists?

Reply:
we(adal/msal) do not validate tokens received from sts

Reply:
true, but you are testing the entire result, I'm thinking that at least one value besides from the upn an be validated. I'm not asking you to look into the tokens.

---
In reply to: [224197546](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1304#discussion_r224197546) [](ancestors = 224197546)